[APM] Implementing Journey for APM

achyutjhunjhunwala opened the pull request:
## Summary

Closes - https://github.com/elastic/kibana/issues/153844

As part of this PR, as its just the stepping stone, we will only cover a basic navigation flow and analyze the result obtained from Steps Dashboard and data collected by the APM Agents for this journey

## Scope

- Generating a data set using Synthtrace instead of Archives
- Capturing the flow from Service Inventory to Trace Waterfall loading on Transaction page
- Capturing Event loop utilisation metrics enabled for APM Journey

## How to run it

```
node scripts/run_performance.js --journey-path x-pack/performance/journeys/apm_service_inventory.ts --skip-warmup
```

achyutjhunjhunwala committed:
Initial setup for APM journey

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

achyutjhunjhunwala committed:
ADD EBT for Service Inventory

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

achyutjhunjhunwala committed:
Fix time capture issue for EBT calculation due to useEffect nature

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

dgieselaar commented on x-pack/plugins/apm/public/components/app/service_inventory/index.tsx:
```diff
@@ -202,8 +204,33 @@ function useServicesDetailedStatisticsFetcher({
+    core: { analytics },
+  } = useApmPluginContext();
+
+  const fetchMainStatisticsStartTime = useRef(0);
```
Can we put this in a separate hook? Or maybe as an option on useFetcher? I think @sqren might have ideas here.

achyutjhunjhunwala replied:
i actually started thinking in that direction that may be we can expose 2 variables with the useFetcher
1. `shouldReportMetrics` which will be default false, and can be enabled.
2. A custom event name, which can be passed to if the above is set to true.

But then that would have required a bit more effort in refactoring a lot of code. So open for idea here

sqren replied:
We discussed this offline and agreed to move the time measuring to the journey itself if possible. 
If that's not possible I suggested to calculate latency for APIs in x-pack/plugins/apm/server/routes/apm_routes/register_apm_server_routes.ts

rylnd committed:
Risk Score Persistence API (#161503)

## Summary

* Introduces a new API, POST `/api/risk_scores/calculate`, that triggers
the code introduced here
* As with the [preview
route](https://github.com/elastic/kibana/pull/155966), this endpoint is
behind the `riskScoringRoutesEnabled` feature flag
* We intend to __REMOVE__ this endpoint before 8.10 release; it's mainly
a convenience/checkpoint for testing the existing code. The next PR will
introduce a scheduled Task Manager task that invokes this code
periodically.
* Updates to the /preview route:
* `data_view_id` is now a required parameter on both endpoints. If a
dataview is not found by that ID, the id is used as the general index
pattern to the query.
* Response has been updated to be more similar to the [ECS risk
fields](https://github.com/elastic/ecs/pull/2236) powering this data.
* Mappings created by the [Data
Client](https://github.com/elastic/kibana/pull/158422) have been updated
to be aligned to the ECS risk fields (linked above)
* Adds/updates the [OpenAPI
spec](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml)
for these endpoints; useful starting point if you're trying to get
oriented here.


## Things to review
* [PR Demo
environment](https://rylnd-pr-161503-risk-score-task-api.kbndev.co/app/home)
* Preview API and related UI still works as expected
* Calculation/Persistence API correctly bootstraps/persists data
    * correct mappings/ILM are created
    * things work in non-default spaces
   



### Checklist


- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ]
[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html)
was added for features that require explanation or tutorials
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to
identify risks that should be tested prior to the change/feature
release.

When forming the risk matrix, consider some of the following examples
and how they may potentially impact the change:

| Risk | Probability | Severity | Mitigation/Notes |

|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space.
| Low | High | Integration tests will verify that all features are still
supported in non-default Kibana Space and when user switches between
spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions
when multiple Kibana nodes are polling for the same tasks. | High | Low
| Tasks are idempotent, so executing them multiple times will not result
in logical error, but will degrade performance. To test for this case we
add plenty of unit tests around this logic and document manual testing
procedure. |
| Code should gracefully handle cases when feature X or plugin Y are
disabled. | Medium | High | Unit tests will verify that any feature flag
or plugin combination still results in our service operational. |
| [See more potential risk
examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled
appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

achyutjhunjhunwala committed:
Remove EBT from APM code

achyutjhunjhunwala committed:
Update journey to cover flow upto trace waterfall

achyutjhunjhunwala commented:
Closing PR as it got a bit messed up while rebasing
